neqo-transport/src/ecn.rs

```diff
@@ -0,0 +1,205 @@
+/// See RFC9000, Appendix A.4.
+#[derive(Debug, PartialEq, Default, Clone)]
+enum EcnValidationState {
+    /// The path is currently being tested for ECN capability.
+    #[default]
+    Testing,
+    /// The validation test has concluded but the path's ECN capability is not yet known.
+    Unknown,
+    /// The path is known to **not** be ECN capable.
+    Failed,
+    /// The path is known to be ECN capable.
+    Capable,
+}
+
+/// The counts for different ECN marks.
+#[derive(PartialEq, Eq, Debug, Clone, Copy, Default)]
+pub struct EcnCount(EnumMap<IpTosEcn, u64>);
+
+impl Deref for EcnCount {
+    type Target = EnumMap<IpTosEcn, u64>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for EcnCount {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl EcnCount {
+    pub fn new(not_ect: u64, ect0: u64, ect1: u64, ce: u64) -> Self {
+        // Yes, the enum array order is different from the argument order.
+        Self(EnumMap::from_array([not_ect, ect1, ect0, ce]))
+    }
+
+    /// Whether any of the ECN counts are non-zero.
+    pub fn is_some(&self) -> bool {
+        self[IpTosEcn::Ect0] > 0 || self[IpTosEcn::Ect1] > 0 || self[IpTosEcn::Ce] > 0
+    }
+}
+
+impl Sub<EcnCount> for EcnCount {
+    type Output = EcnCount;
+
+    /// Subtract the ECN counts in `other` from `self`.
+    fn sub(self, other: EcnCount) -> EcnCount {
+        let mut diff = EcnCount::default();
+        for (ecn, count) in &mut *diff {
+            *count = self[ecn].saturating_sub(other[ecn]);
+        }
+        diff
+    }
+}
+
+impl AddAssign<IpTosEcn> for EcnCount {
+    fn add_assign(&mut self, ecn: IpTosEcn) {
+        self[ecn] += 1;
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct EcnInfo {
+    /// The current state of ECN validation on this path.
+    state: EcnValidationState,
+
+    /// The number of probes sent so far on the path during the ECN validation.
+    probes_sent: usize,
+
+    /// The largest ACK seen so far.
+    largest_acked: PacketNumber,
+
+    /// The ECN counts from the last ACK frame that increased `largest_acked`.
+    baseline: EcnCount,
+}
+
+impl EcnInfo {
+    /// Set the baseline (= the ECN counts from the last ACK Frame).
+    pub fn set_baseline(&mut self, baseline: EcnCount) {
+        self.baseline = baseline;
+    }
+
+    /// Expose the current baseline.
+    pub fn baseline(&self) -> EcnCount {
+        self.baseline
+    }
+
+    /// Count the number of packets sent out on this path during ECN validation.
+    /// Exit ECN validation if the number of packets sent exceeds `ECN_TEST_COUNT`.
+    /// We do not implement the part of the RFC that says to exit ECN validation if the time since
+    /// the start of ECN validation exceeds 3 * PTO, since this seems to happen much too quickly.
+    pub fn on_packet_sent(&mut self) {
+        if self.state != EcnValidationState::Testing {
+            return;
+        }
```

Comment:
You could move the probe count to the state:

```suggestion
        if let EcnValidationState::Testing(probes_sent) = &mut self.state else {
            return;
        }
        *probes_sent += 1;
```

That doesn't make sense for `baseline` or `largest_acked`, but it's a possibility here.